// @sentry/electron/dist/dispatch.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The Sentry Electron SDK Frontend.
 *
 * @see ElectronOptions for documentation on configuration options.
 * @see SentryClient for usage documentation.
 */
class ElectronFrontend {
    /**
     * Creates a new Electron SDK instance.
     *
     * This constructor automatically chooses the right implementation for the
     * process type (`browser` or `renderer`).
     *
     * @param options Configuration options for this SDK.
     */
    constructor(options) {
        // We dynamically load the frontend implementation for the current process
        // type. In frontend bundlers such as webpack or rollup, those requires are
        // resolved statically. For this reason, we use `module.require` for the
        // main implementation here, which is only defined in the main process. The
        // renderer implementation must use the default `require`.
        // In case `process.type` is not defined, dispatch defaults to the renderer
        // implementation, which should be fine for most cases. False positives of
        // this would be running `@sentry/electron` in a bare node process, which is
        // acceptable.
        // tslint:disable:no-var-requires no-unsafe-any
        const frontendClass = process.type === 'browser'
            ? module.require('./main').MainFrontend
            : require('./renderer').RendererFrontend;
        // tslint:enable:no-var-requires no-unsafe-any
        this.inner = new frontendClass(options);
    }
    /**
     * @inheritDoc
     */
    captureMinidump(path, event, scope) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.inner.captureMinidump(path, event, scope);
        });
    }
    /**
     * @inheritDoc
     */
    install() {
        return this.inner.install();
    }
    /**
     * @inheritDoc
     */
    captureException(exception, scope) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.inner.captureException(exception, scope);
        });
    }
    /**
     * @inheritDoc
     */
    captureMessage(message, scope) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.inner.captureMessage(message, scope);
        });
    }
    /**
     * @inheritDoc
     */
    captureEvent(event, scope) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.inner.captureEvent(event, scope);
        });
    }
    /**
     * @inheritDoc
     */
    addBreadcrumb(breadcrumb, scope) {
        this.inner.addBreadcrumb(breadcrumb, scope);
    }
    /**
     * @inheritDoc
     */
    getDSN() {
        return this.inner.getDSN();
    }
    /**
     * @inheritDoc
     */
    getOptions() {
        return this.inner.getOptions();
    }
    /**
     * @inheritDoc
     */
    setContext(context, scope) {
        this.inner.setContext(context, scope);
    }
    /**
     * @inheritDoc
     */
    getInitialScope() {
        return this.inner.getInitialScope();
    }
}
exports.ElectronFrontend = ElectronFrontend;
